pkg/ctlog/config.go

```diff
@@ -0,0 +1,324 @@
+// Copyright 2022 The Sigstore Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ctlog
+
+// This module contains helpers for manipulating CTLog configuration. Since the
+// configuration is in a protobuf and we have to marshal/unmarshal, and update
+// it during the operation of the CTLog / Fulcio, hoisted into here for easier
+// testing.
+
+import (
+	"bytes"
+	"context"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"strings"
+
+	"github.com/google/certificate-transparency-go/trillian/ctfe/configpb"
+	"github.com/google/trillian/crypto/keyspb"
+	"github.com/sigstore/sigstore/pkg/cryptoutils"
+	"google.golang.org/protobuf/encoding/prototext"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/anypb"
+	"knative.dev/pkg/logging"
+)
+
+const (
+	// Key in the configmap holding the value of the tree.
+	treeKey = "treeID"
+	// ConfigKey is the key in the map holding the marshalled CTLog config.
+	ConfigKey = "config"
+	// PrivateKey is the key in the map holding the encrypted PEM private key
+	// for CTLog.
+	PrivateKey = "private"
+	// PublicKey is the key in the map holding the PEM public key for CTLog.
+	PublicKey = "public"
+	// LegacyRootCAKey is the key for when we only supported a single entry
+	// in the config.
+	LegacyRootCAKey = "rootca"
+	bitSize         = 4096
+
+	// This is hardcoded since this is where we mount the certs in the
+	// container.
+	rootsPemFileDir = "/ctfe-keys/"
+	// This file contains the private key for the CTLog
+	privateKeyFile = "/ctfe-keys/privkey.pem"
+)
+
+// CTLogConfig abstracts the proto munging to/from bytes suitable for working
+// with secrets / configmaps. Note that we keep fulcioCerts here though
+// technically they are not part of the config, however because we create a
+// secret/CM that we then mount, they need to be synced.
+type CTLogConfig struct {
+	PrivKey         *rsa.PrivateKey
```

Comment:
Can we use an ECDSA key instead? It'll produce smaller signatures over SCTs, and that will result in smaller certificates since SCTs are embedded in issued certs.